### ISSUE TITLE
Generated meson files fail to build

### ISSUE THREAD

**Issue by cromerc** [rastersoft/autovala#84]
I am trying to generate meson files for our project that has AutoVala in use: https://github.com/cromnix/pamac-classic

 The CMake files work great and everything builds just fine. However the meson files don't work. This is the output I get when trying to build:
```
meson --prefix=/usr --sysconfdir=/etc
The Meson build system
Version: 0.43.0
Source dir: /home/cromer/Desktop/projects/pamac-classic/test
Build dir: /home/cromer/Desktop/projects/pamac-classic/test/build
Build type: native build

Meson encountered an error in file meson.build, line 252, column 0:
Assignment target must be an id.
cfg_pamac-installer = configuration_data()
^

A full log can be found at /home/cromer/Desktop/projects/pamac-classic/test/build/meson-logs/meson-log.txt
```

This is the generated meson.build file: https://gist.github.com/cromerc/38782d85b179410905382e5a0e8cb440

**Comment by rastersoft** [comment]
Mmm... I suspect that the problem could be because that name has a hyphen in its name...

**Comment by rastersoft** [comment]
I will do a change to remove the hyphens... give a couple of hours...

**Comment by rastersoft** [comment]
Try with this file: http://www.rastersoft.com/meson.build

**Comment by cromerc** [comment]
Well it got further this time, but crashed on a different line.
```
meson --prefix=/usr --sysconfdir=/etc
The Meson build system
Version: 0.43.0
Source dir: /home/cromer/Desktop/projects/pamac-classic/test
Build dir: /home/cromer/Desktop/projects/pamac-classic/test/build
Build type: native build
Project name: pamac
Native C compiler: ccache cc (gcc 7.2.0)
Native Vala compiler: valac (valac 0.38.2)
Build machine cpu family: x86_64
Build machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (0.29.2)
Native dependency gdk-3.0 found: YES 3.22.24
Native dependency gtk+-3.0 found: YES 3.22.24
Native dependency json-glib-1.0 found: YES 1.4.2
Native dependency libalpm found: YES 10.0.2
Native dependency libcurl found: YES 7.56.0
Native dependency libnotify found: YES 0.7.7
Native dependency libsoup-2.4 found: YES 2.60.0
Native dependency polkit-gobject-1 found: YES 0.114
Native dependency vte-2.91 found: YES 0.50.1
Native dependency gio-2.0 found: YES 2.54.0
Native dependency glib-2.0 found: YES 2.54.0
Native dependency gobject-2.0 found: YES 2.54.0
Program /home/cromer/Desktop/projects/pamac-classic/test/meson_scripts/install_data.sh found: YES (/home/cromer/Desktop/projects/pamac-classic/test/meson_scripts/install_data.sh)
Program glib-compile-resources found: YES (/usr/bin/glib-compile-resources)
Program glib-compile-resources found: YES (/usr/bin/glib-compile-resources)
Program glib-compile-resources found: YES (/usr/bin/glib-compile-resources)
Configuring Config_1.vala using configuration

Meson encountered an error in file meson.build, line 156, column 0:
Unknown variable "pamac_manager_no_aur_gresource_xml_file_c".

A full log can be found at /home/cromer/Desktop/projects/pamac-classic/test/build/meson-logs/meson-log.txt
```

**Comment by rastersoft** [comment]
Ok, it is a problem because that is defined with a conditional... let me check it...

**Comment by cromerc** [comment]
Just gave it a try, getting much further, but still not a success. Running meson worked just fine now. But now when running ninja it fails to build with this output:
```
[24/92] Compiling C object 'AUR@sha/meson-generated_Config_2.c.o'.
FAILED: AUR@sha/meson-generated_Config_2.c.o 
ccache cc  -IAUR@sha -I. -I.. -I/usr/include/json-glib-1.0 -I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include -I/usr/include/libsoup-2.4 -I/usr/include/libxml2 -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -w -O0 -g '-DGETTEXT_PACKAGE="pamac"' -fPIC -pthread -MMD -MQ 'AUR@sha/meson-generated_Config_2.c.o' -MF 'AUR@sha/meson-generated_Config_2.c.o.d' -o 'AUR@sha/meson-generated_Config_2.c.o' -c 'AUR@sha/Config_2.c'
cc: error: AUR@sha/Config_2.c: No such file or directory
cc: fatal error: no input files
compilation terminated.
[28/92] Compiling Vala source Config_1...la ../src/transaction_sum_dialog.vala.
ninja: build stopped: subcommand failed.
```

**Comment by rastersoft** [comment]
No idea... And my debian system lacks libalpm, so I can't try it :(

**Comment by rastersoft** [comment]
Anyway, don't forget that the project uses the "include" command in autovala, and that is not portable to meson... Maybe the problem is there...

**Comment by cromerc** [comment]
Those include shouldn't effect anything for meson, those include I have are just to fix the order in which the binaries/libraries are built under cmake. For example: https://github.com/cromnix/pamac-classic/blob/master/src/pamac-tray/CMakeDepends.txt

I just added add_dependencies(pamac-tray Pamac AUR) This makes sure those 2 libraries are built before it tries to build the pamac-tray binary.
It is a bit hacky, but by doing this I got parallel build working. Now I don't have to do make -j1 I can use all 8 cores. Technically it would be nice if AutoVala did this automatically, but I got it to work with this simple hack.

Yeah, unfortunately libalpm which is included in the pacman package under arch linux is not available for debian. It would require you to build and install it manually, however it doesn't have any special dependencies which means it can be built using all the libraries that are already available under debian. Or of course a virtual machine with arch linux would work too. ;)

Is there anything I can do on my end to help you with debugging this?

**Comment by rastersoft** [comment]
Ok, so that's the problem with parallel build... I have to try to fix that.

**Comment by rastersoft** [comment]
I will try to find that library and test it again... but tomorrow. Now it is 23:42 in spain and I'm really tired O:)

**Comment by cromerc** [comment]
No problem, thanks for all the help. :)

**Comment by rastersoft** [comment]
I tried to compile libalpm, but with no success. I had to download the pacman repository, and when trying to run autogen.sh, it fails.

**Comment by rastersoft** [comment]
Ok, Arch in virtual machine...

**Comment by rastersoft** [comment]
BTW autovala adds the "add_dependencies" in theory.

**Comment by cromerc** [comment]
Then something is wrong, the add_dependencies are not there which is why I added them manually.

**Comment by rastersoft** [comment]
No, sorry, the "add_dependencies" are added only for gresources files. My fault.

**Comment by rastersoft** [comment]
Hi again:
I added support for automatic dependencies, so now the code in the master branch should be able to add the "add_dependencies" entries in cmake. I did some tests, but can you test it by removing the includes and compiling it with cmake?

If it works fine, I will add the same to meson. I suspect that that will fix your last error.

**Comment by cromerc** [comment]
@rastersoft, I just gave it a try and it is not done correctly. All of the add_dependencies are added in the main CMakeLists.txt file that is in the at the root of the project. For it to work correctly with parallel builds each add_dependencies must be in the CMakeLists.txt file that is in the executable/library directory.
Here is some info on this on the cmake mailing list:
https://cmake.org/pipermail/cmake/2008-January/019021.html
As noted, they say that the add_dependencies must be done in the same folder as the binary/library.

**Comment by cromerc** [comment]
I also just noticed, but it looks like the conditions are wrong in the code generated as well:
```
add_dependencies( pamac-install Pamac )
add_dependencies( pamac-manager Pamac )
add_dependencies( pamac-system-daemon Pamac if (NOT DISABLE_AUR)
	AUR )
endif ()
add_dependencies( pamac-tray Pamac if (NOT DISABLE_AUR)
	AUR )
endif ()
add_dependencies( pamac-user-daemon AUR Pamac )
```
The output is malformed and fails to build. Here is the log output:
```
CMake Error at CMakeLists.txt:87 (endif):
  endif An ENDIF command was found outside of a proper IF ENDIF structure.
  Or its arguments did not match the opening IF command.


CMake Error at CMakeLists.txt:90 (endif):
  endif An ENDIF command was found outside of a proper IF ENDIF structure.
  Or its arguments did not match the opening IF command.


-- Configuring incomplete, errors occurred!
See also "/home/cromer/Desktop/projects/pamac-classic/build/CMakeFiles/CMakeOutput.log".
```

**Comment by rastersoft** [comment]
Ops!!!!! Ok, fixing them! Give me some minutes...

**Comment by rastersoft** [comment]
Ok, now both problems should be fixed. Please, try it again, if you don't mind.

**Comment by cromerc** [comment]
Working good now. Parallel build was successful. All that is left is the meson problem.

**Comment by rastersoft** [comment]
Great! I'll check it tomorrow, today is 1:36 am O:)

**Comment by cromerc** [comment]
Me parece bien, gracias.

**Comment by rastersoft** [comment]
Ah, claro, que eres de Chile :D

**Comment by cromerc** [comment]
I finally got meson to build after modifying a lot of things, and found some bugs in autovala's meson generation.

I was able to solve some of the build errors by creating meson.build files in the subdirectories of each binary/library. By doing that it fixed the Config_X.c missing errors. I had to change some meson.current_source_dir() calls to meson.source_root() in the new meson.build files in the subdirs. I then removed src/ from the source lists since now they look in the local folder for the source files.

Now that I have meson.build files in the subdirs I got much further, but ran into a bug in autovala:
```
[30/86] Linking target src/libPamac.so.6.0.1.
FAILED: src/libPamac.so.6.0.1 
ccache cc  -o src/libPamac.so.6.0.1 'src/Pamac@sha/meson-generated_pamac.installer.gresource.xml.c.o' 'src/Pamac@sha/meson-generated_pamac.manager.gresource.xml.c.o' 'src/Pamac@sha/meson-generated_pamac.transaction.gresource.xml.c.o' 'src/Pamac@sha/meson-generated_Config_1.c.o' 'src/Pamac@sha/meson-generated_alpm_config.c.o' 'src/Pamac@sha/meson-generated_choose_ignorepkgs_dialog.c.o' 'src/Pamac@sha/meson-generated_choose_provider_dialog.c.o' 'src/Pamac@sha/meson-generated_common.c.o' 'src/Pamac@sha/meson-generated_mirrors_config.c.o' 'src/Pamac@sha/meson-generated_package.c.o' 'src/Pamac@sha/meson-generated_pamac_config.c.o' 'src/Pamac@sha/meson-generated_preferences_dialog.c.o' 'src/Pamac@sha/meson-generated_progress_box.c.o' 'src/Pamac@sha/meson-generated_transaction.c.o' 'src/Pamac@sha/meson-generated_transaction_sum_dialog.c.o' -Wl,--no-undefined -Wl,--as-needed -shared -fPIC -Wl,--start-group -Wl,-soname,libPamac.so.6 -lgdk-3 -lpangocairo-1.0 -lpango-1.0 -lgdk_pixbuf-2.0 -lcairo-gobject -lcairo -lgobject-2.0 -lglib-2.0 -Wl,--end-group -lgtk-3 -lgdk-3 -lpangocairo-1.0 -lpango-1.0 -latk-1.0 -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lalpm -lnotify -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lvte-2.91 -lgtk-3 -lgdk-3 -lpangocairo-1.0 -lpango-1.0 -latk-1.0 -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lz -lpcre2-8 -lgnutls -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lglib-2.0 -lgobject-2.0 -lglib-2.0 '-Wl,-rpath,$ORIGIN/' -Wl,-rpath-link,/home/cromer/Desktop/projects/pamac-classic/build/src  
src/Pamac@sha/meson-generated_pamac.manager.gresource.xml.c.o: In function `pamac_get_resource':
/home/cromer/Desktop/projects/pamac-classic/build/src/Pamac@sha/pamac.manager.gresource.xml.c:5544: multiple definition of `pamac_get_resource'
src/Pamac@sha/meson-generated_pamac.installer.gresource.xml.c.o:/home/cromer/Desktop/projects/pamac-classic/build/src/Pamac@sha/pamac.installer.gresource.xml.c:324: first defined here
src/Pamac@sha/meson-generated_pamac.transaction.gresource.xml.c.o: In function `pamac_get_resource':
/home/cromer/Desktop/projects/pamac-classic/build/src/Pamac@sha/pamac.transaction.gresource.xml.c:5045: multiple definition of `pamac_get_resource'
src/Pamac@sha/meson-generated_pamac.installer.gresource.xml.c.o:/home/cromer/Desktop/projects/pamac-classic/build/src/Pamac@sha/pamac.installer.gresource.xml.c:324: first defined here
src/Pamac@sha/meson-generated_transaction.c.o: In function `pamac_transaction_on_emit_download':
/home/cromer/Desktop/projects/pamac-classic/src/transaction.vala:1520: undefined reference to `ceilf'
/home/cromer/Desktop/projects/pamac-classic/src/transaction.vala:1522: undefined reference to `ceilf'
/home/cromer/Desktop/projects/pamac-classic/src/transaction.vala:1568: undefined reference to `ceilf'
/home/cromer/Desktop/projects/pamac-classic/src/transaction.vala:1570: undefined reference to `ceilf'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```
It is failing here because in the meson.build file generated by autovala it is including the wrong gresource files with the binaries/libraries.

For example in my pamac.avprj file, I have this:
```
vala_library: src/pamac
version: 6.0.1
*namespace: Pamac
use_gresource: pamac_transaction_gresource_xml
vala_package: posix
vala_check_package: gdk-3.0
vala_check_package: gtk+-3.0
vala_check_package: libalpm
vala_check_package: libnotify
vala_check_package: vte-2.91
*vala_check_package: gio-2.0
*vala_check_package: glib-2.0
*vala_check_package: gobject-2.0
*vala_source: alpm_config.vala
*vala_source: choose_ignorepkgs_dialog.vala
*vala_source: choose_provider_dialog.vala
*vala_source: common.vala
*vala_source: mirrors_config.vala
*vala_source: package.vala
*vala_source: pamac_config.vala
*vala_source: preferences_dialog.vala
*vala_source: progress_box.vala
*vala_source: transaction.vala
*vala_source: transaction_sum_dialog.vala
```
There is only 1 gresource file being used, however in my meson.build generated by autvala this is the output:
```
pamac_sources += [pamac_installer_gresource_xml_file_c]
if DISABLE_AUR
pamac_sources += [pamac_manager_no_aur_gresource_xml_file_c]
pamac_sources += [pamac_transaction_no_aur_gresource_xml_file_c]
else
pamac_sources += [pamac_manager_gresource_xml_file_c]
pamac_sources += [pamac_transaction_gresource_xml_file_c]
endif
pamac_vala_args = ['--pkg','posix']
pamac_vala_args += ['--vapidir='+join_paths(meson.current_source_dir(),'src/vapis')]
if DISABLE_AUR
pamac_vala_args += ['--gresources='+join_paths(meson.current_source_dir(),'data/pamac.transaction_no_aur.gresource.xml')]
else
pamac_vala_args += ['--gresources='+join_paths(meson.current_source_dir(),'data/pamac.transaction.gresource.xml')]
endif
```
As you can see the meson.build file has other gresource files listed that are not actually used. What I noticed is that all binaries and all libraries list all the gresources. For some reason autovala is adding all of them to every binary and every library.

I had to modify [meson_scripts/install_library.sh](https://github.com/cromnix/pamac-classic/blob/master/meson_scripts/install_library.sh)
I added the directory "src" and a third argument $3. This is to be able to handle the build of each subdir I wrote. Then in the subdir libraries I added the 3rd argument so that it enters into the correct directory.

I removed all the extra stuff from the root meson.build and added:
```
subdir('src')
subdir('src/aur')
subdir('src/pamac-install')
subdir('src/pamac-manager')
subdir('src/pamac-system-daemon')
subdir('src/pamac-user-daemon')
subdir('src/pamac-tray')
```
In some of the binaries I link against 2 of the libraries generated by this project. So I needed to add this to the meson.build file:
```
inc = include_directories(['../', '../aur'])
executable('pamac-system-daemon',pamac_system_daemon_sources,dependencies: pamac_system_daemon_deps,vala_args: pamac_system_daemon_vala_args,c_args: pamac_system_daemon_c_args,link_with: pamac_system_daemon_dependencies,install: true,include_directories : inc)
```
This allows the header files to be found, without doing this the build would error out saying the library header file could not be find. It has to be relative paths, doing this does not work with full paths.

Also by doing subdir builds I could change:
`cfgfile_5 = configure_file(input: 'Config.vala.base',output: 'Config_1.vala',configuration: cfg_pamac_system_daemon)`
to this:
`cfgfile_5 = configure_file(input: 'Config.vala.base',output: 'Config.vala',configuration: cfg_pamac_system_daemon)`
The number is not necessary because the file is in a different folder.

Here is the commit where I fixed the meson.build files: https://github.com/cromnix/pamac-classic/commit/0db40030c2e7ef0025df673a0c7b73603c725f05

Hopefully this info will help you to correct autovala's meson support and get it working. All I did was manually modify the meson.build files to get it to work which means that if I run autovala meson it would stop working again.

**Comment by cromerc** [comment]
Vivo en Chile hace 7 años pero soy estadounidense.

Chris Cromer

**Comment by rastersoft** [comment]
Great! With that I think I'll be able to fix it.

Also I'm doing some changes to the dependencies detection: I'm adding a new configuration statement, "compile_dependency" to allow to check which dependencies are being added automatically by autovala, and to manually add others when the automatic detection can't determine all the ones needed.

**Comment by rastersoft** [comment]
I've nearly ended to do the needed changes, but I have two questions:

- In "src/aur/meson.build" you set TESTSRCDIR to meson.source_root(), but in all other "meson.build" files it is set to "meson.current_source_dir()". Which one is the correct one?

- In the executables, you specify "include_directories: '../'. Why do you specify that, and why?

Thanks.

**Comment by cromerc** [comment]
- TESTSRCDIR should be source_root, I forgot to change it in the other files. The "tests" for a project should be at the root of the project in general, so it should be source_root.
- Because the main part of the project is a library not a binary.
/src - library
/src/aur - library
/src/pamac-* - binary
When I need to link to the main library I need "../", when I need the aur library I need "../aur".
basically the include_directories needs to be set to the relative paths to the libraries needed. If you do not specify that, the build will fail saying that the libraries header files cannot be found.

**Comment by rastersoft** [comment]
Ok... working on it.

**Comment by rastersoft** [comment]
Ok, good news: everything works now in meson, tested in Debian and... Tadah!!!! In Arch linux.. And also fixed some extra bugs.

**Comment by cromerc** [comment]
Great! I will try it out tomorrow.

**Comment by cromerc** [comment]
Works great, everything bulids and packages fine. The only remaining issue before I can adopt meson officially in the project is how to handle symlinks.
Right now I have an include which runs cmake code to create the symlink during install. Only problem is that the include doesn't help me with meson.
The best solution will be symlink support in autovala: https://github.com/rastersoft/autovala/issues/85

Anyway, I will go ahead and close this issue since meson now builds.

**Comment by rastersoft** [comment]
BTW: the parallel compilation should work fine now, so you should be able to remove those external cmake files.

**Comment by cromerc** [comment]
Thanks for noticing, I will remove those now as well.
